### PR TITLE
Add Erlang OTP Application cataloger

### PR DESCRIPTION
**Pull request by LaurentGoderre**
Adds a cataloger for Erlang OTP applications.

Depends on:
- #2390 
- https://github.com/anchore/packageurl-go/pull/16

ErLang OTP Manifest format: https://www.erlang.org/doc/man/app.html
OTP Purl proposal: https://erlef.github.io/security-wg/specs/otp_purl_type.html

### CODE REVIEW

File: syft/internal/packagemetadata/names.go [hunk 1]
```diff
@@ -93,6 +93,7 @@ var jsonTypes = makeJSONTypes(
 	jsonNames(pkg.PythonPipfileLockEntry{}, "python-pipfile-lock-entry", "PythonPipfileLockMetadata"),
 	jsonNames(pkg.PythonRequirementsEntry{}, "python-pip-requirements-entry", "PythonRequirementsMetadata"),
 	jsonNames(pkg.ErlangRebarLockEntry{}, "erlang-rebar-lock-entry", "RebarLockMetadataType"),
+	jsonNames(pkg.ErlangOTPApplication{}, "erlang-otp-application", "OTPApplicationMetadataType"),
```
[the remark this answers is not in the document]
Comment (LaurentGoderre) [reply syft/internal/packagemetadata/names.go:96]: If I remove it, the tests start failing:

```
--- FAIL: TestReflectTypeFromJSONName_LegacyValues (0.00s)
    --- FAIL: TestReflectTypeFromJSONName_LegacyValues/map_pkg.ErlangOTPApplication_struct_type (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0xb8 pc=0x101050148]

goroutine 86 [running]:
testing.tRunner.func1.2({0x1011f4f80, 0x101744c70})
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1545 +0x1c4
testing.tRunner.func1()
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1548 +0x360
panic({0x1011f4f80?, 0x101744c70?})
        /opt/homebrew/opt/go/libexec/src/runtime/panic.go:914 +0x218
github.com/anchore/syft/syft/internal/packagemetadata.TestReflectTypeFromJSONName_LegacyValues.func1(0x14000642680?)
        /Users/laurentgoderre/git/scanner/syft/syft/internal/packagemetadata/names_test.go:259 +0xb8
testing.tRunner(0x14000642820, 0x140006966b0)
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 44
        /opt/homebrew/opt/go/libexec/src/testing/testing.go:1648 +0x33c
FAIL    github.com/anchore/syft/syft/internal/packagemetadata   0.890s
```

File: go.mod [hunk 2]
```diff
@@ -239,3 +239,5 @@ retract (
 	v0.53.2
 	v0.53.1 // Published accidentally with incorrect license in depdencies
 )
+
+replace github.com/anchore/packageurl-go => github.com/LaurentGoderre/packageurl-go v0.0.0-20231207160803-decab5854c64
```
Comment (LaurentGoderre) [inline go.mod:243]: Only here for the tests to pass. Remove when https://github.com/anchore/packageurl-go/pull/16 lands

File: syft/pkg/erlang.go [hunk 3]
```diff
@@ -7,3 +7,8 @@ type ErlangRebarLockEntry struct {
 	PkgHash    string `mapstructure:"pkgHash" json:"pkgHash"`
 	PkgHashExt string `mapstructure:"pkgHashExt" json:"pkgHashExt"`
 }
+
+type ErlangOTPApplication struct {
```
Comment (wagoodman) [inline syft/pkg/erlang.go:11]: the metadata type isn't really needed if it is exclusively for name and version (which is already on package)

**Comment by wagoodman** [comment]
I can spend a little time today getting this rebased and ready for review/merge

File: go.mod (hunk 2 above)
Comment (wagoodman) [reply go.mod]: I've removed the replace and used the raw value and left a comment. We try to keep as many replace directives out of the go.mod as possible.

**Review by wagoodman — APPROVED** [review]

**Pull request merged**